src/json_schema/mod.rs

```diff
@@ -1,84 +1,1101 @@
+
+        let json_value: Value = serde_json::from_str(json).expect("Can't parse json");
+        let regex = to_regex(&json_value, None);
+        assert!(regex.is_ok(), "{:?}", regex);
+
+        // Confirm the depth of 3 recursion levels by default, recursion level starts
+        // when children start to have children
+        let re = Regex::new(&regex.unwrap()).expect("Regex failed");
+        for lvl in [
+            // level 0
+            r#"{ "name": "Az"}"#,
+            r#"{ "name": "Az", "children": [] }"#,
+            r#"{ "name": "Az", "children": [{"name": "Bo"}] }"#,
+            // level 1
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [] }] }"#,
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li"}] }] }"#,
+            // level 2
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li", "children": [] }] }] }"#,
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li", "children": [{"name": "Ho"}] }] }] }"#,
+            // level 3
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li", "children": [{"name": "Ho", "children": [] }] }] }] }"#,
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li", "children": [{"name": "Ho", "children": [{"name": "Ro"}] }] }] }] }"#,
+        ] {
+            should_match(&re, lvl);
+        }
+
+        for lvl in [
+            // level 4
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li", "children": [{"name": "Ho", "children": [{"name": "Ro", "children": [] }] }] }] }] }"#,
+            r#"{ "name": "Az", "children": [{"name": "Bo", "children": [{"name": "Li", "children": [{"name": "Ho", "children": [{"name": "Ro", "children": [{"name": "Ks"}] }] }] }] }] }"#,
+        ] {
+            should_not_match(&re, lvl);
         }
-        _ => Err(anyhow!("Invalid JSON Schema: expected an object")),
+    }
+
+    #[test]
+    fn indirect_recursion_with_recursion_level_regex_match() {
+        let json = r##"{
+          "type": "object",
+          "properties": {
+              "node": { "$ref": "#/definitions/node" }
+          },
+          "definitions": {
+              "node": {
+                  "type": "object",
+                  "properties": {
+                      "value": { "type": "integer" },
+                      "next": { "$ref": "#/definitions/node" }
+                  }
+              }
+          }
+        }"##;
+        let json_value: Value = serde_json::from_str(json).expect("Can't parse json");
+        let mut parser = parsing::Parser::new(&json_value).with_max_recursion_depth(0);
+
+        let result = parser.to_regex(&json_value);
+        assert!(result.is_ok(), "{:?}", result);
+        let regex = result.unwrap();
+        assert_eq!(
+            r#"\{([ ]?"node"[ ]?:[ ]?\{([ ]?"value"[ ]?:[ ]?(-)?(0|[1-9][0-9]*))?[ ]?\})?[ ]?\}"#,
```

Comment:
I think this is fine for generation, but a thing to document might be that a string representing a valid json object might not match this because of the 0-or-1 whitespace option. Also are you avoiding `\w` on purpose?

Reply:
Whitespace pattern behavior wasn't touched, it's exactly as it was before. In this case if whitespace pattern isn't defined, so default one is used. It's a lot of work to re-write all the tests with proper aliases like `[WHITESPACE]` and python tests gave up half way to support it via aliases, so as I did in related ported ones in rust, which was a lot on its own. 

If one day default whitespace will change it would be easier to call the logic with defined whitespace of ` ` than to update all I guess.

Reply:
That makes a lot of sense! Mostly just wanted to flag for the future!